### PR TITLE
[Feature] : add mm input item cache support

### PR DESCRIPTION
**Pull request by kousakawang**
## Modifications
related to issue: https://github.com/sgl-project/sglang/issues/8433

### Image Input Data Cache & CUDA IPC Support

This feature introduces **image input caching** and **CUDA IPC-based transmission** for the `qwen2-image-processor` and `glm4v-image-processor`.  
It is designed to reduce redundant computation and minimize inter-process communication overhead.

#### 🔄 Image Cache
- If an image has already been processed, it will be **retrieved from cache** instead of being processed again.
- Avoids repeated preprocessing of the same image.
- Stored tensors remain in the `image_processor`’s internal cache table.

#### ⚡ CUDA IPC Transmission
- Replaces the default **raw tensor transmission via ZMQ** with a **CUDA IPC handle dictionary**.
- Reduces transmission time for large `pixel_values` tensors.
- **Important:** CUDA IPC is only used when caching is enabled, ensuring that cached tensors are not recycled by PyTorch’s memory manager.

### Use Cases

- **AI Agents**  
  Multiple images are often passed as input, but usually only one is new. Caching prevents redundant preprocessing of repeated images.

- **Robotics / Navigation**  
  Frequent re-use of similar frames can benefit from fast cache retrieval and IPC-based transmission.

### How to Enable

```bash
export SGL_CACHE_MM_IMAGE=1
# set cache max tokenizer cache size,  or will use default value or max-available memorysize)
export SGL_TOKENIZER_CACHED_IMAGE_SIZE_MB=1024 #(max cache 1024MB image data)

```

### [TODO]
~~mange cache by using MB_SIZE instead of  image nums~~ (done)

### Notes

* Works with both qwen2-image-processor and glm4v-image-processor.

* CUDA IPC is automatically applied only in cache mode.

* Designed for scenarios where image reuse is common, improving both speed and efficiency.

### Ut for this Feature
python3 -m unittest test_vision_openai_server_a.TestQwen2_5_AI_AGENT.test_ai_agent_opt

### test_result: 

**Accuray: (no change with Feature On/Off)**

Action_Summary: yes, I understand. firstly, I need to open the browser
Action: click(start_box="<bbox>367 7   

Action_Summary: yes, I understand. firstly, I need to open the browser
Action: click(start_box="<bbox>367 7

**Performance:**
opt : 709.1031 ms  (with  cache)
normal:  2222.0058  ms (wo cache)

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

## Summary by CodeRabbit

- New Features
  - Added multimodal image caching for GLM-4v and Qwen VL processors with async support and env-based controls.
  - Enabled handling of dict-based image features during batching.
- Performance
  - Faster repeated image requests via GPU-backed cache and lightweight image hashing, reducing latency and memory overhead.
- Bug Fixes
  - Improved stability by safely handling optional model attributes to prevent runtime errors.
- Tests
  - Introduced end-to-end tests comparing cached vs. non-cached flows, validating identical outputs and measurable latency improvements.

### CODE REVIEW

File: python/sglang/srt/managers/schedule_batch.py [hunk 1]
```diff
@@ -1270,6 +1270,40 @@ def prepare_for_extend(self):
...
+
+                    real_pixel_values = []
+                    take_start_idx = 0
+                    take_end_idx = 0
+                    take_idx = 0
+                    for send_idx in range(len(use_cache_marks)):
+                        if use_cache_marks[send_idx]:
+                            real_pixel_values.append(
+                                self.mminput_cache[hash_keys[send_idx]]
+                            )
```
[the remark this answers is not in the document]
Comment (kousakawang) [reply python/sglang/srt/managers/schedule_batch.py]: done

File: python/sglang/srt/multimodal/processors/base_processor.py [hunk 2]
```diff
@@ -223,18 +325,179 @@ def process_mm_data(
...
+                            device=input_ids.device,
+                        )
+                        new_tensor[0][:pos] = input_ids[0][:pos]
+                        new_tensor[0][pos : pos + len(insert_ids)] = torch.tensor(
+                            insert_ids, dtype=input_ids.dtype, device=input_ids.device
+                        )
+                        new_tensor[0][pos + len(insert_ids) :] = input_ids[0][pos + 1 :]
+                        return new_tensor
+
+                return input_ids
```
[the remark this answers is not in the document]
Comment (kousakawang) [reply python/sglang/srt/multimodal/processors/base_processor.py]: Done

File: python/sglang/srt/multimodal/processors/base_processor.py [hunk 3]
```diff
@@ -9,11 +10,110 @@
...
+
+
+def image_to_int(img: Image.Image) -> int:
+    img_bytes = img.tobytes()
+    hash_obj = hashlib.md5(img_bytes)
+    hash_bytes = hash_obj.digest()
+    return int.from_bytes(hash_bytes, byteorder="big")
+
+
+class FIFOHashTable:
```
[the remark this answers is not in the document]
Comment (kousakawang) [reply python/sglang/srt/multimodal/processors/base_processor.py]: Done

File: python/sglang/srt/multimodal/processors/base_processor.py [hunk 4]
```diff
@@ -223,18 +325,179 @@ def process_mm_data(
...
+                            insert_ids, dtype=input_ids.dtype, device=input_ids.device
+                        )
+                        new_tensor[0][pos + len(insert_ids) :] = input_ids[0][pos + 1 :]
+                        return new_tensor
+
+                return input_ids
+
+            to_replace_str = "<|vision_start|><|image_pad|><|vision_end|>"
+            repalce_str = "<|vision_end|>"
+            v_start_token, img_pad_token, v_end_token = 151652, 151655, 151653
```
[the remark this answers is not in the document]
Comment (kousakawang) [reply python/sglang/srt/multimodal/processors/base_processor.py]: done

File: python/sglang/srt/multimodal/processors/base_processor.py [hunk 5]
```diff
@@ -233,18 +362,152 @@ def process_mm_data(
...
+                        ]
+                    )
+                    use_cache_mark.append(False)
+                # add input ids and insert tensor
+                else:
+                    cached_tensor = self.hash_table.get(img_hash_keys[img_idx])
+                    assert isinstance(
+                        cached_tensor, torch.Tensor
+                    ), "invalid cached_tensor"
+                    # tensor_lists.append(cached_tensor)
```
Comment (byjiang1996) [inline python/sglang/srt/multimodal/processors/base_processor.py]: Is there any reason why `self.hash_table's` value is not used but eventually use the value from `mminput_cache`?

It seems `self.hash_table` alone would suffice

File: python/sglang/srt/managers/schedule_batch.py [hunk 6]
```diff
@@ -1290,6 +1292,40 @@ def prepare_for_extend(self):
                 pixel_values = getattr(mm_item, "feature", None)
                 if isinstance(pixel_values, torch.Tensor):
                     mm_item.feature = pixel_values.to(self.device, non_blocking=True)
+                elif isinstance(pixel_values, dict):
```
Comment (byjiang1996) [inline python/sglang/srt/managers/schedule_batch.py:1236]: Is there any specific reason why this logic cannot be moved over into base_processor.py?

File: python/sglang/srt/multimodal/processors/base_processor.py [hunk 7]
```diff
@@ -233,18 +362,152 @@ def process_mm_data(
...
+                img_token_num = int(img_height / QWEN2_PER_TOKEN_PATCH_NUM)
+
+                if img_token_num < 1:
+                    raise ValueError("invalid img token num")
+
+                if self.hash_table.get(hash_key) is None:
+                    new_processed_img_idxes.append(img_idx)
+                    new_processed_imgs.append(images[img_idx])
+                else:
+                    remove_image_idx.append(img_idx)
```
Comment (byjiang1996) [inline python/sglang/srt/multimodal/processors/base_processor.py]: Is there any specific reason why cache is used only once? (not sure if i understood correctly)

File: python/sglang/srt/multimodal/processors/base_processor.py [hunk 8]
```diff
@@ -233,18 +362,152 @@ def process_mm_data(
...
+            # result["pixel_values"] = torch.concat(tensor_lists, dim = 0)
+            send_pixel_values = (
+                torch.concat(tensor_lists, dim=0) if len(tensor_lists) > 0 else None
+            )
+            proxy_pixel_values = {}
+            proxy_pixel_values["send_data"] = send_pixel_values
+            proxy_pixel_values["use_cache_mark"] = use_cache_mark
+            proxy_pixel_values["hash_keys"] = img_hash_keys
+            proxy_pixel_values["img_heights"] = processed_img_heights
+            proxy_pixel_values["delete_keys"] = delete_keys
```
Comment (byjiang1996) [inline python/sglang/srt/multimodal/processors/base_processor.py]: Similar to this [comment](https://github.com/sgl-project/sglang/pull/8742/files#r2284122037), I'm wondering why we need `proxy_pixel_values` but can't set `pixel_values` correctly within base_processor.py

File: python/sglang/srt/multimodal/processors/base_processor.py [hunk 9]
```diff
@@ -233,18 +362,152 @@ def process_mm_data(
...
+                # add input ids and insert tensor
+                else:
+                    cached_tensor = self.hash_table.get(img_hash_keys[img_idx])
+                    assert isinstance(
+                        cached_tensor, torch.Tensor
+                    ), "invalid cached_tensor"
+                    # tensor_lists.append(cached_tensor)
+                    insert_cached_ids = (
+                        [v_start_token]
+                        + img_token_nums[img_idx] * [img_pad_token]
```
Comment (byjiang1996) [inline python/sglang/srt/multimodal/processors/base_processor.py]: I'm wondering if there is any `result['xx']` field that can be used directly to obtain img_token_num instead of calculating `img_token_num` manually

File: python/sglang/srt/multimodal/processors/base_processor.py [hunk 10]
```diff
@@ -233,18 +362,152 @@ def process_mm_data(
...
+            delete_keys = []
+            for img_idx in range(len(images)):
+                # cache Tensor
+                if img_idx in new_processed_img_idxes:
+                    img_height = img_heights[img_idx]
+                    processed_img_heights.append(img_height)
+             
+                    start_height = end_height
+                    end_height = start_height + img_height
+                    to_cache_tensor = result["pixel_values"][start_height:end_height]
```
Comment (byjiang1996) [inline python/sglang/srt/multimodal/processors/base_processor.py]: Similar comments as https://github.com/sgl-project/sglang/pull/8742/files#r2284154152

Is there any result['xx'] field to use instead of calculating `img_height/start_height/end_height` manually?

File: python/sglang/srt/multimodal/processors/base_processor.py [hunk 11]
```diff
@@ -233,18 +362,152 @@ def process_mm_data(
...
+            transformers.models.qwen2_vl.image_processing_qwen2_vl_fast.Qwen2VLImageProcessorFast,
         )
-        # move feature tensors to cpu
-        for feature_name in self.FEATURE_NAMES:
-            if feature_name in result and isinstance(
-                result[feature_name], torch.Tensor
-            ):
-                result[feature_name] = result[feature_name].to("cpu")
+
+        if cache_mm_image_items and is_qwen2_processor:
```
Comment (byjiang1996) [inline python/sglang/srt/multimodal/processors/base_processor.py]: Is this logic universal to most VLMs? (seems not from my limited understanding)

If no, could you please move this logic into qwen vl processor's processor adapter similar to what we did for [phi4-mm](https://github.com/sgl-project/sglang/blob/7e8187e00428d292b98526ba7b78f95a30f05fba/python/sglang/srt/multimodal/processors/phi4mm.py#L18)? That way this model specfic code can be moved over to model.py instead of in common code places

File: python/sglang/srt/multimodal/processors/base_processor.py [hunk 12]
```diff
@@ -9,11 +10,137 @@
...
+
+
+def image_to_int(img: Image.Image) -> int:
+    img_bytes = img.tobytes()
+    hash_obj = hashlib.md5(img_bytes)
+    hash_bytes = hash_obj.digest()
+    return int.from_bytes(hash_bytes, byteorder="big")
+
+
+class FIFOHashTable:
```
Comment (byjiang1996) [inline python/sglang/srt/multimodal/processors/base_processor.py]: Similar comments as https://github.com/sgl-project/sglang/pull/8742/files#r2284122880

If we can use cache more than once, we should consider `@lru_cache` i think

File: python/sglang/srt/managers/schedule_batch.py (hunk 6 above)
Comment (kousakawang) [reply python/sglang/srt/managers/schedule_batch.py:1236]: I think in base_processor.py, we just load the image (raw or base64) and convert it to tensor(which will be accepted by model as input) , this action happen in tokenizer process.   After this  tokenizer process will send data to scheduler process by using ZMQ. As a result, if we only cache data in tokenizer process, the scheduler can not get the correct pixel values.

File: python/sglang/srt/multimodal/processors/base_processor.py (hunk 11 above)
Comment (kousakawang) [reply python/sglang/srt/multimodal/processors/base_processor.py]: yes, you are right, this  logic just for Qwen2VLImageProcessorFast,  I will try to move code to qwen2.

File: python/sglang/srt/multimodal/processors/base_processor.py (hunk 7 above)
Comment (kousakawang) [reply python/sglang/srt/multimodal/processors/base_processor.py]: Not sure I undrerstand your question correctly.  If you mean why do “remove_image_idx.append(img_idx)” here， this is for rocording which image was alredy cached, and then the coresponding image token  will be removed from input text to avoiding processsing image data err(ths input_txt should hold  same image nums which is equal to len(new_processed_imgs)

File: python/sglang/srt/multimodal/processors/base_processor.py (hunk 10 above)
Comment (kousakawang) [reply python/sglang/srt/multimodal/processors/base_processor.py]: yes, I think we can store the height of each image in to a struct and next time we can use it directly. But for an new image , still we need to calculate it. just because each image hold different size and we can not use a  fixed value. (The more imporant reason is that , we can not modify the logic in  "processor.__call__" which was  realized in transformers not sglang...)

File: python/sglang/srt/multimodal/processors/base_processor.py (hunk 5 above)
Comment (kousakawang) [reply python/sglang/srt/multimodal/processors/base_processor.py]: As this comment  https://github.com/sgl-project/sglang/pull/8742/files#r2284939106 said: 
mminput_cache was manged by scheduler subprocess,
and self.hash_table was managed by tokenizer subprocess.
Inter-process communication is done through ICP (each process uses a different port) via the ZMQ library.
so we cache the image data in tokenizer process  aims at :  

1. avoiding processing the same image again, 
2. and also saving the data-send cost between tokenizer process and scheduler process.

To avoid sending a pixel-values(torch.tensor) again, we need to established mm_cache in scheduler process, also. (and this can be a gpu tensor cache)
By doing so , for a normal image, the normal processing route is GPU-> cpu ->send via ZMQ -> cpu ->GPU
for a cached image, we can process it  via raw_image or bas64 -> send hash -> getGPU tensor to avoid send any real pixel values or do any D2H H2D.

Actually, I think the IPC-handle is a better way to avoid sending data but share gpu memory between differnt process, ( in this case, cache is not necessary)
But since get ipc-handle will not increase the ref count of torch's tensor, so without keeping all data always active (by a global container or someting) , access tensor via ipc-handle in scheduler process will have some risk.  (also, ipchandle can only avoiding the send cost, but not the image processing cost)

File: python/sglang/srt/multimodal/processors/base_processor.py (hunk 8 above)
Comment (kousakawang) [reply python/sglang/srt/multimodal/processors/base_processor.py]: This is because we can not modify the result["pixel_value"] which is processed via transformer's processor.

**Comment by kousakawang** [comment]
close since cuda-ipc feature  was merged in #11917
